explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,252 @@
+
+The first thing we have to do is to do several runs of `openssl speed` *before* enabling PGO, so that we have a baseline to compare with.  As explained in the sections above, `autofdo` does not handle stripped binaries nor `dwz`-processed debug information, so we need to make sure the resulting binary obeys these restrictions.  See the section below for more details on that.
+
+After confirming that everything looks OK, we are ready to start the benchmark.  Let's run the command:
+
+```bash
+$ openssl speed -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+This will run benchmark tests for `md5`, `sha512`, `rsa2048` and `aes-256-cbc`.  Each test will last 60 seconds, and they will involve calculating as many cryptographic hashes from `N`-byte chunks of data (with `N` varying from 16 to 16384) with each algorithm (with the exception of `rsa2048`, whose performance is measured in signatures/verifications per second).  By the end, you should see a report like the following:
+
+```bash
+The 'numbers' are in 1000s of bytes per second processed.
+type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
+sha512           91713.74k   366632.96k   684349.30k  1060512.22k  1260359.00k  1273277.92k
+aes-256-cbc    1266018.36k  1437870.33k  1442743.14k  1449933.84k  1453336.84k  1453484.99k
+md5             140978.58k   367562.53k   714094.14k   955267.21k  1060969.81k  1068296.87k
+                  sign    verify    sign/s verify/s
+rsa 2048 bits 0.000173s 0.000012s   5776.8  86791.9
+```
+
+### Building OpenSSL for profiling
+
+Before we are able to profile OpenSSL, we need sure that we compile the software in a way that the generated program meets the requirements.  In our case, you can use the `file` command after building the binary to make sure that it has not been stripped:
+
+```bash
+$ file /usr/bin/openssl
+/usr/bin/openssl: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=42a5ac797b6981bd45a9ece3f67646edded56bdb, for GNU/Linux 3.2.0, with debug_info, not stripped
+```
+
+Note how `file` reports the binary as being `with debug_info, not stripped`.  This is the output you need to look for when inspecting your own binaries.
+
+You can use the `eu-readelf` command (from the `elfutils` package) to determine whether any compression was performed with `dwz`:
+
+```bash
+$ eu-readelf -S /usr/bin/openssl
+[...]
+Section Headers:
+[Nr] Name                 Type         Addr             Off      Size     ES Flags Lk Inf Al
+[...]
+[31] .debug_abbrev        PROGBITS     0000000000000000 001bcbf3 000110c4  0        0   0  1
+[...]
+```
+
+If the `Flags` field has `C` (as in `compressed`) in it, this means that `dwz` was used.  You will then need to figure out how to disable its execution during the build.
```

Comment:
It's as easy as not invoking `dwz` during the build.  It's just hard to write a generic way to do that without knowing the particularities of the build system being used.

Reply:
Maybe thereby lead in the second sentence with "Since the particularities how to skip that depend on the build system you are using [You will then need ...]"?

If you WANT to go deeper you might say - "For my build this was done via ... "

Reply:
Done.